File: java/leetcode_grind/Day424.java
package leetcode_grind;

import java.util.Stack;

public class Day424 {
    // https://leetcode.com/problems/amount-of-time-for-binary-tree-to-be-infected/description
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    static class Solution1 {
        int maxDistance = 0;

        public int amountOfTime(TreeNode root, int start) {
            traverse(root, start);
            return maxDistance;
        }

        int traverse(TreeNode root, int start) {
            int depth = 0;
            if (root == null) {
                return depth;
            }

            int leftDepth = traverse(root.left, start); // 2
            int rightDepth = traverse(root.right, start); // -1

            if (root.val == start) {
                maxDistance = Math.max(leftDepth, rightDepth);
                depth = -1;
            } else if (leftDepth >= 0 && rightDepth >= 0) {
                depth = Math.max(leftDepth, rightDepth) + 1;
            } else {
                int distance = Math.abs(leftDepth) + Math.abs(rightDepth);
                maxDistance = Math.max(maxDistance, distance);
                depth = Math.min(leftDepth, rightDepth) - 1;
            }

            return depth;
        }
    }

    // https://leetcode.com/problems/remove-k-digits/description/
    static class Solution2 {
        public String removeKdigits(String num, int k) {
            var stack = new Stack<Character>();
            for (int i = 0; i < num.length(); i++) {
                while (!stack.isEmpty() && k > 0 && stack.peek() > num.charAt(i)) {
                    stack.pop();
                    k--;
                }
                stack.push(num.charAt(i));
            }
            while (k-- > 0) {
                stack.pop();
            }
            var sb = new StringBuilder();
            for (var n : stack) {
                if (sb.isEmpty() && n == '0') {
                    continue;
                }
                sb.append(n);
            }
            return sb.isEmpty() ? "0" : sb.toString();
        }
    }
}

File: python/leetcode_grind/day_424.py
# https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/description/
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        def add_parent(cur, parent):
            if cur:
                cur.parent = parent
                add_parent(cur.left, cur)
                add_parent(cur.right, cur)

        add_parent(root, None)

        answer = []
        visited = set()

        def dfs(cur, distance):
            if not cur or cur in visited:
                return

            visited.add(cur)

            if distance == 0:
                answer.append(cur.val)
                return

            dfs(cur.parent, distance - 1)
            dfs(cur.left, distance - 1)
            dfs(cur.right, distance - 1)

        dfs(target, k)
        return answer

#User function Template for python3

class Solution:

    def removeKdigits(self, S, K):
        # code here
        st = []
        for ch in S:
            while len(st) > 0 and K > 0 and st[-1] > ch:
                st.pop()
                K -= 1

            if not len(st) and ch == '0':
                continue
            st.append(ch)
        
        while len(st) and K > 0:
            st.pop()
            K -= 1
        
        return "0" if len(st) == 0 else "".join(st)
        
